File: android/app/src/main/java/edu/illinois/rokwire/poll/PollPlugin.java
/*
 * Copyright 2020 Board of Trustees of the University of Illinois.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package edu.illinois.rokwire.poll;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import edu.illinois.rokwire.MainActivity;
import edu.illinois.rokwire.Utils;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

public class PollPlugin implements MethodChannel.MethodCallHandler {

    private static final String TAG = "PollPlugin";

    private MainActivity context;
    private MethodChannel methodChannel;

    private PollBleServer blePollServer;
    private PollBleClient bleClient;

    public static PollPlugin registerWith(PluginRegistry.Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "edu.illinois.rokwire/polls");
        PollPlugin pollPlugin = new PollPlugin((MainActivity)registrar.activity(), channel);
        channel.setMethodCallHandler(pollPlugin);
        return pollPlugin;
    }

    private PollPlugin(MainActivity activity, MethodChannel methodChannel) {
        this.context = activity;
        this.methodChannel = methodChannel;
        this.methodChannel.setMethodCallHandler(this);

        bindPollServer();
        bindPollClient();
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
        try {
            if ("create_poll".equals(call.method)) {
                if (blePollServer != null) {
                    Object arguments = call.arguments;
                    if (arguments instanceof String) {
                        String pollID = (String) arguments;
                        blePollServer.createPoll(pollID);
                    } else {
                        Log.d(TAG, "arguments is null or not in the correct type");
                    }
                }
            } else if ("start_scan".equals(call.method)) {
                if (bleClient != null) {
                    bleClient.startScan();
                }
            }
            else if ("stop_scan".equals(call.method)) {
                if (bleClient != null) {
                    bleClient.stopScan();
                }
            }
            else if("enable".equals(call.method)){
                // Implement if need
                if (bleClient != null) {
                    bleClient.startScan();
                }
            }
            else if("disable".equals(call.method)){
                // Implement if need
                if (bleClient != null) {
                    bleClient.stopScan();
                }
            }
            result.success(null);
        } catch (IllegalStateException exception) {
            String errorMsg = String.format("Ignoring exception '%s'. See https://github.com/flutter/flutter/issues/29092 for details.", exception.toString());
            Log.e(TAG, errorMsg);
            exception.printStackTrace();
        }
    }

    public void onLocationPermissionGranted() {
        Log.d(TAG, "onLocationPermissionGranted");
        if (bleClient != null)
            bleClient.onLocationPermissionGranted();
    }

    private void requestBluetoothOn() {
        Log.d(TAG, "requestBluetoothOn");

        Utils.showDialog(context, "Illinois app",
                "Using bluetooth nearby devices can receive your poll. Do you want to turn the bluetooth on? ",
                (dialog, which) -> {
                    //Turn bluetooth on
                    Utils.enabledBluetooth();

                }, "Yes",
                (dialog, which) -> {}, "No",
                true);
    }

    // Poll server

    private void bindPollServer() {
        Intent intent = new Intent(context, PollBleServer.class);
        context.bindService(intent, mConnection, Context.BIND_AUTO_CREATE);
    }

    private ServiceConnection mConnection = new ServiceConnection() {
        public void onServiceConnected(ComponentName className, IBinder service) {
            blePollServer = ((PollBleServer.LocalBinder)service).getService();
            if (blePollServer == null) {
                return;
            }
            blePollServer.setCallback(serverCallback);
        }
        public void onServiceDisconnected(ComponentName className) {
            blePollServer = null;
        }
    };

    private PollBleServer.Callback serverCallback = new PollBleServer.Callback() {
        @Override
        public void onRequestBluetoothOn() {
            PollPlugin.this.requestBluetoothOn();
        }
    };

    /////////////////////

    // Poll client

    private void bindPollClient() {
        Intent intent = new Intent(context, PollBleClient.class);
        context.bindService(intent, clientConnection, Context.BIND_AUTO_CREATE);
    }

    private ServiceConnection clientConnection = new ServiceConnection() {
        public void onServiceConnected(ComponentName className, IBinder service) {
            bleClient = ((PollBleClient.LocalBinder)service).getService();
            if (bleClient == null) {
                return;
            }
            bleClient.setCallback(clientCallback);
        }
        public void onServiceDisconnected(ComponentName className) {
            bleClient = null;
        }
    };

    private PollBleClient.Callback clientCallback = new PollBleClient.Callback() {
        @Override
        public void onPollCreated(String pollId) {
            Log.d(TAG, "onPollCreated - " + pollId);

            //run it on the ui thread
            Handler handler = new Handler(Looper.getMainLooper());
            handler.post(() -> methodChannel.invokeMethod("on_poll_created", pollId));
        }
    };

    /////////////////////
}

File: ios/Runner/PollPlugin.m
//
//  PollServer.m
//  Runner
//
//  Created by Mladen Dryankov on 13.12.19.
//  Copyright 2020 Board of Trustees of the University of Illinois.
    
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#import "PollPlugin.h"
#import "AppDelegate.h"
#import "NSDictionary+InaTypedValue.h"
#import "NSString+InaJson.h"
#import "NSData+InaHex.h"
#import "Bluetooth+InaUtils.h"

#import <CoreBluetooth/CoreBluetooth.h>
#import <UserNotifications/UserNotifications.h>

// Channel Commands
static NSString* const ChannelStartScan	 							= @"start_scan";
static NSString* const ChannelStopScan	 							= @"stop_scan";
static NSString* const ChannelCreatePoll	 						= @"create_poll";

static NSString* const ChannelEnable		 						= @"enable";
static NSString* const ChannelDisable		 						= @"disable";

// Notification
static NSString* const ChannelOnPollCreated	 						= @"on_poll_created";

// Periphere Name
static NSString* const PollPluginPeriphereName		 				= @"Poll";

@interface PollPlugin()<CBPeripheralManagerDelegate, CBCentralManagerDelegate>{
	bool 					initialized;
	
	FlutterMethodChannel	*channel;
	
	CBCentralManager		*centralManager;
	CBPeripheralManager		*peripheralManager;
	
	NSMutableSet			*processedPollIds;
	
	NSTimer					*timoutTimer;
}
@property(nonatomic,assign,readonly) bool 		isAuthorized;
@end

@implementation PollPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar{
	FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:@"edu.illinois.rokwire/polls" binaryMessenger:registrar.messenger];
	PollPlugin *instance = [[PollPlugin alloc] initWithChannel:channel];
	[registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)init
{
	if (self = [super init]) {
		processedPollIds = [NSMutableSet new];
	}
	return self;
}

- (instancetype)initWithChannel:(FlutterMethodChannel*)_channel
{
	if (self = [self init]) {
		channel = _channel;
	}
	return self;
}

-(void)initialize{
	if(self.isAuthorized && !initialized){
		peripheralManager = [[CBPeripheralManager alloc] initWithDelegate:self queue:nil];
		centralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil options:@{
			CBCentralManagerOptionShowPowerAlertKey : [NSNumber numberWithBool:NO],
		}];
		
		initialized = true;
	}
}

-(void)deinitialize{
	if(initialized){
		[self stopScan];
		peripheralManager.delegate = nil;
		peripheralManager = nil;
		centralManager.delegate = nil;
		centralManager = nil;
		initialized = false;
	}
}

-(bool)isAuthorized{
	return (InaBluetooth.peripheralAuthorizationStatus == InaBluetoothAuthorizationStatusAuthorized) && (InaBluetooth.centralAuthorizationStatus == InaBluetoothAuthorizationStatusAuthorized);
}

//////////////
// Server Side

- (void)startWithPollId:(NSString*)_pollId{
	[self initialize];
	
	// Dont start server if it's not preperly initialized
	if(!initialized || !self.isAuthorized){
		NSLog(@"Poll service is not properly setup, initialized or authorized");
		return;
	}
	
	CBUUID *pollUuid = [self cbUuidFromString:_pollId];

	if(peripheralManager.isAdvertising){
		[peripheralManager stopAdvertising];
	}
	
	if(pollUuid != nil){
		[peripheralManager startAdvertising:
			@{ CBAdvertisementDataServiceUUIDsKey :@[pollUuid],
			   CBAdvertisementDataLocalNameKey:PollPluginPeriphereName
		}];
		[self startTimoutTimer];
	}
}

- (void)stop{
	
	// Dont stop server if it's not preperly initialized at the beginning
	if(!initialized){
		NSLog(@"Poll service is not properly setup && initialized");
		return;
	}
	[self stopTimeoutTimer];
	[peripheralManager stopAdvertising];
}

#pragma mark CBPeripheralManager

-(void)peripheralManagerDidUpdateState:(CBPeripheralManager *)peripheral{
	NSLog(@"CBPeripheralManager State: %@",@(peripheral.state));
}

-(void)startTimoutTimer{
	[self stopTimeoutTimer];
	timoutTimer = [NSTimer scheduledTimerWithTimeInterval:5*60 repeats:NO block:^(NSTimer * _Nonnull timer) {
		[self stop];
	}];
}

-(void)stopTimeoutTimer{
	if(timoutTimer != nil){
		[timoutTimer invalidate];
		timoutTimer = nil;
	}
}

//////////////
// Client Side

-(void)startScan{
	if (centralManager.state == CBManagerStatePoweredOn) {
		[centralManager scanForPeripheralsWithServices:nil options:@{ CBCentralManagerScanOptionAllowDuplicatesKey : @YES }];
	}
}

-(void)stopScan{
	[centralManager stopScan];
}

#pragma mark CBCentralManagerDelegate

- (void)centralManagerDidUpdateState:(CBCentralManager *)central{
	if (centralManager.state == CBManagerStatePoweredOn) {
		[self startScan];
	}
	else{
		[self stopScan];
	}
}

- (void)centralManager:(CBCentralManager *)central didDiscoverPeripheral:(CBPeripheral *)peripheral advertisementData:(NSDictionary<NSString *, id> *)advertisementData RSSI:(NSNumber *)RSSI{
	
	// With this we ensure it's a valid UUID otherwise ignore
	NSString *periphereName = [advertisementData inaStringForKey:CBAdvertisementDataLocalNameKey];
	NSArray *serviceUuids = [advertisementData inaArrayForKey:CBAdvertisementDataServiceUUIDsKey];
	if([PollPluginPeriphereName isEqualToString:periphereName] && serviceUuids.count > 0){
		NSString *convertedId = [self pollIdFromCbUuid:serviceUuids.firstObject];
		
		if(![processedPollIds containsObject:convertedId]){
			[channel invokeMethod:ChannelOnPollCreated arguments:convertedId];
			[processedPollIds addObject:convertedId];
		}
	}
	
}

#pragma mark Utils

- (NSString*) pollIdFromCbUuid:(CBUUID*)uuid{
	if(uuid != nil){
		NSData *uuidData = [uuid data];
		NSData *convertedData = uuidData.length >= 12 ? [uuidData subdataWithRange:NSMakeRange(0, 12)] : nil;
		NSString *convertedId = [convertedData inaHexString];
		return convertedId;
	}
	return nil;
}

- (CBUUID*) cbUuidFromString:(NSString*)string{
	if(string != nil){
		NSMutableData *pollIdData = [NSMutableData dataWithData:string.inaDataFromHex];
		if(pollIdData.length < 16){
			[pollIdData increaseLengthBy:(16 - pollIdData.length)];
			return [CBUUID UUIDWithData:pollIdData];
		}
		else{
			return [CBUUID UUIDWithData:[pollIdData subdataWithRange:NSMakeRange(0, 16)]];
		}
	}
	return nil;
}

#pragma mark MethodCall

-(void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result{
	if([ChannelStartScan isEqualToString:call.method]){
		[self startScan];
	}
	else if([ChannelStopScan isEqualToString:call.method]){
		[self stopScan];
	}
	else if([ChannelCreatePoll isEqualToString:call.method]){
		if([call.arguments isKindOfClass:NSString.class]){
			NSString *value = (NSString*) call.arguments;
			if(value.length > 0){
				[self startWithPollId:value];
			}
		}
	}
	else if([ChannelEnable isEqualToString:call.method]){
		[self initialize];
	}
	else if([ChannelDisable isEqualToString:call.method]){
		[self deinitialize];
	}
	result(nil);
}
@end
